### ISSUE TITLE
Implement callback interfaces for Swift

### ISSUE THREAD

**Issue by jhugman** [mozilla/uniffi-rs#353]
This came out of #342 ; but now covers just the work in Swift.

[Callback interfaces][1]:

```
callback interface Filesystem {
    void write_text(string filepath, string text);
    string read_text(string filepath);
};

namespace my_lib {
    set_filesystem(Filesystem fs);
}
```

[1]: https://heycam.github.io/webidl/#dfn-callback-interface

**Comment by jhugman** [comment]
This should be a much smaller issue once #352 is implemented.

**Comment by DAOCUONG** [comment]
@jhugman  i would like to help with this issue can you give me guidance how can i start

**Comment by jhugman** [comment]
Hi @DAOCUONG !

Thanks for showing an interest in uniffi.

There are quite a lot of moving parts to this feature; it's quite well documented in https://github.com/mozilla/uniffi-rs/pull/344.

I'd start by working out what the Swift version of `ForeignCallback` is: I expect you'll need to add something to the `Template-Bridging-Header.h`. To start with, don't worry about arguments or return values. 

Make a swift implementation of that `ForeignCallback`, and find a way to pass it to rust, and prove to yourself that you rust can now call swift. All this will be fairly hacky to start with, but without this, nothing else matters. Let me know when you've got here, because we can start talking next steps.

[1]: https://github.com/mozilla/uniffi-rs/pull/344/files#diff-ea603f09fea8bf67e546b39042d59df87507d0ec5b8660dd2f464c30ff305ca6

**Comment by patrick-fitzgerald** [comment]
Hi @jhugman, 

I'm interested in helping with this feature. 

I'm currently using Kotlin callbacks in an Android app using uniffi-rs, I'd like to also use callbacks in an iOS app with Swift.

Would this project be a good reference for the implementation of Swift callbacks? 
https://github.com/xajik/rust-cross-platform-mobile/blob/master/swapi_ios/src/ffi/callback.rs

Thanks

**Comment by rfk** [comment]
Hi @patrick-fitzgerald, thanks for your interest! I'm not familiar with the linked project, but from a first glance it looks like it may be taking a bit of a different approach than we take in UniFFI, plumbing Rust and Swift together directly rather than going via a shared C-level FFI layer.

There are a couple of layers of things to figure out at the Swift level. I recommend setting yourself up to work through them by hand in the first instance - it's a bit fiddly but it helps iterate faster. Something along the lines of:

* Go into the `./examples/callbacks` directory and poke around a bit to understand the example.
* Copy the structure of its `tests/bindings/test_callbacks.kts` file to make a `test_callbacks.swift` and sketch out the code you'd expect to write in there if this feature was working today. Add it to the list in `./tests/test_generated_bindings.rs`.
    * You could start with just the simplest function call with no args or return values, then progressive add more of the bindings as you get things working.
* Run `cargo test swift` in the callbacks example directory. It'll fail, probably with a panic telling you that Swift callbacks are not implemented. Track down the source of this panic in `uniffi_bindgen` and change it to spit out some placeholder code like "// TODO".
* Run `cargo test swift` again. This time it should generate the Swift bindings, then fail when it tries to run the actual test. In the `../../target/debug/` directory you will get:
    * `build/uniffi-example-callbacks-[some sort of hash]/out/callbacks.uniffi.rs`: the generated Rust FFI layer.
    * `uniffi_callbacks-Bridging-Header.h`: this defines the low-level FFI as exposed by the Rust code.
    * `uniffi_callbacks.modulemap`: this tells swift how to import the low-level FFI as a Swift module.
    * `callbacks.swift`: these are the high-level Swift bindings.
* Edit the test runner so that it won't overwrite these files, by commenting out [this line](https://github.com/mozilla/uniffi-rs/blob/d44b96c9de105a6fafc4ff1e1f66e3248631a129/uniffi_bindgen/src/lib.rs#L205). Make a change to `uniffi_callbacks-Bridging-Header.h`, run `cargo test swift` again, and confirm that it doesn't overwrite your changes.

Having typed that all out, it sounds like a lot of messing around and we should probably figure out a way to smooth that iteration loop a bit! But still, I think it's worth getting set up to try fleshing out the bindings by hand while keeping a nice testing loop via `cargo test`.

From there, you'll need to figure out what to put in the `.h` file to expose the callback machinery into swift. Basically you want to turn the `pub extern "C"` API generated by [`CallbackInterfaceTemplate.rs`](https://github.com/mozilla/uniffi-rs/blob/d44b96c9de105a6fafc4ff1e1f66e3248631a129/uniffi_bindgen/src/scaffolding/templates/CallbackInterfaceTemplate.rs#L1) into some C header definitions. I don't think there will be anything Swift-specific here, it will just be some functions that take function pointers as arguments in ordinary C style.

Then, you'll need to figure out how to consume that API from within `callbacks.swift`. I have no expertise here, but it seems like there's pretty good support for this in Swift, e.g. [this article](http://www.perry.cz/clanky/swift.html) and some of the notes on [this page](https://developer.apple.com/documentation/swift/swift_standard_library/manual_memory_management/calling_functions_with_pointer_parameters). The project you linked above might have useful context here too!

Basically, that becomes an iterative process of figuring out the code to put in those two files until `cargo test swift` passes to your satisfaction. Once it does, it should be "just" a matter of templating the generation of that code by adding it into the [Swift code templates](https://github.com/mozilla/uniffi-rs/tree/main/uniffi_bindgen/src/bindings/swift/templates).

I do understand that that's a *lot* to tackle here. But if you're keen to dive in, I'm happy to answer questions or review work-in-progress PRs or anything else I can do to help you along that road.

**Comment by patrick-fitzgerald** [comment]
Hi @rfk, thanks for the detailed walkthrough of the changes I'd need to make! 

I'm going to be working part-time on this over the coming weeks. I have a WIP PR on my fork that I'll push my code to: https://github.com/patrick-fitzgerald/uniffi-rs/pull/1 - until it's ready for a PR on this repo. 
I'll let you know if I have any Q's. Thanks!

**Comment by rfk** [comment]
> `uniffi_callbacks-Bridging-Header.h`: this defines the low-level FFI as exposed by the Rust code.
> `uniffi_callbacks.modulemap`: this tells swift how to import the low-level FFI as a Swift module.

FYI the names of these files have changed recently, they will now be `callbacksFFI.h` and `callbacksFFI.modulemap`.

**Comment by patrick-fitzgerald** [comment]
Hi @rfk, sorry for the delay. I got a chance to look into this a bit further today with the steps you outlined above. It does look like there is a lot to tackle. 😅 
I was testing the implementation in Kotlin and I see it generates the following logic:

```
internal class CallbackInterfaceTestCallbackFFI : ForeignCallback {
    @Suppress("TooGenericExceptionCaught")
    override fun invoke(handle: Long, method: Int, args: RustBuffer.ByValue): RustBuffer.ByValue {
        return CallbackInterfaceTestCallbackFFI.handleMap.callWithResult(handle) { cb ->
            when (method) {
                IDX_CALLBACK_FREE -> CallbackInterfaceTestCallbackFFI.drop(handle)
                1 -> this.invokeOnSuccess(cb, args)
                2 -> this.invokeOnError(cb, args)

                // This should never happen, because an out of bounds method index won't
                // ever be used. Once we can catch errors, we should return an InternalError.
                // https://github.com/mozilla/uniffi-rs/issues/351
                else -> RustBuffer.ByValue()
            }
        }
    }
```

I'd assume we'd use a similar type of approach with Swift? 
I'm not too familiar with C, I might need some help setting a basic implementation of the header in `callbacksFFI.h`.

**Comment by patrick-fitzgerald** [comment]
@rfk Would the callback implementation by the `rc_log_adapter_create` method in the `mozilla/application-services` project be a good example to follow?
https://github.com/mozilla/application-services/search?q=rc_log_adapter_create 
It's implemented on iOS and Android.

**Comment by jhugman** [comment]
@patrick-fitzgerald that would be a great example to follow!

The initial kotlin implementation uses the same jna Callback as is used in RustLog. I think that stores the adaptor on the Rust side (RustLogState stores this on the foreign language side).

For context: `rc_log_adapter_create` is called at library startup to initialize the Rust to Swift channel. Until this is established Rust has no way of calling back out to Swift.

**Comment by patrick-fitzgerald** [comment]
I've been using the `rc_log_adapter_create` method as an example to test callbacks: 

`examples/callbacks/src/callbacks.udl`
```
namespace callbacks {};

interface LibRustLogAdapter {
  void rc_log_adapter_create(RustLogCallback callback);
};

callback interface RustLogCallback {
  void invoke(u64 level, string tag, string msg);
};
```

`examples/callbacks/src/lib.rs:`
```
pub trait RustLogCallback {
    fn invoke(&self, text: u64, tag: String, msg: String);
}


#[derive(Debug, Clone)]
struct LibRustLogAdapter;
impl LibRustLogAdapter {
    fn new() -> Self {
        LibRustLogAdapter
    }
    fn rc_log_adapter_create(&self, callback: Box<dyn RustLogCallback>) {
        callback.invoke(10, "tag".into(), "msg".into());
    }
}

include!(concat!(env!("OUT_DIR"), "/callbacks.uniffi.rs"));
```
`/target/debug/callbacksFFI.h`

```
typedef struct ForeignCallback
{

} ForeignCallback;
```


In `/target/debug/callbacksFFI.swift` I added:
```
public protocol RustLogCallback {
    func invoke(level: Int32, tag: String, msg: String)
}
```

I'm seeing an error:
```
uniffi-rs/target/debug/callbacks.swift:406:83: error: value of type 'RustLogCallback' has no member 'lower'
    callbacks_507c_LibRustLogAdapter_rc_log_adapter_create(self.pointer, callback.lower() , $0
```

I believe the next step would be creating a RustLogCallback extension to convert the callback to a UInt64 with the `callback.lower()` method. Is there any similar example I could use for this functionality? Thanks

**Comment by pagr** [comment]
I needed callbacks in swift for a project I was working on and have done quite some work on this task. It is mostly a copy of how it is done for Kotlin, except that reference counting is only implemented for callbacks and not for classes. I didn't need that part and I'm not sure if it is needed in swift. I have opened a [pull request](https://github.com/mozilla/uniffi-rs/pull/1045) where we can discuss the details.
